### PR TITLE
Enhance ABA-based FD computation by adding motor dynamics

### PR DESCRIPTION
**Pull request by flferretti**
This pull request introduces the possibility for the simulator to accept motor inertia, motor viscous friction and gear ratio, which will be saved in the already existing JAX dataclasses:

- [`high_level.joint`](https://github.com/ami-iit/jaxsim/blob/main/src/jaxsim/high_level/joint.py)
- [`high_level.model`](https://github.com/ami-iit/jaxsim/blob/main/src/jaxsim/high_level/model.py)
- [`model.physics_model`](https://github.com/ami-iit/jaxsim/blob/main/src/jaxsim/physics/model/physics_model.py)

to be ready for Forward Dynamics computation through a new ABA version which incorporates these motor parameters into the simulation process.

As of now, the PR is marked as a draft since even though it successfully passes the tests against [iDynTree](https://github.com/robotology/idyntree), it requires further validation with comprehensive motor dynamics to ensure its robustness and accuracy.

C.C. @diegoferigo @CarlottaSartore @S-Dafarra @traversaro

### CODE REVIEW

**Comment by traversaro** [comment]
Do you have some reference for the math implemented?

**Comment by fl-ferr** [comment]
The problem formalization and the related pseudocode can be found at https://github.com/ami-iit/element_rl-for-codesign/issues/61#issuecomment-1658609058

**Comment by traversaro** [comment]
> The title refers to the fact that this ABA implementation takes into account motor inertias, viscous friction and transmission ratio. I can change the title to "Enhance ABA-based FD computation by considering motor parameters" if that makes things more clear

No problem, once you clarify it in the description I think it is ok!

**Comment by diegoferigo** [comment]
xref of this interesting comment full of references https://github.com/ROAM-Lab-ND/spatial_v2_extended/discussions/6#discussioncomment-6691627, we might have a deeper look at it

**Comment by flferretti** [comment]
Changing the base branch to `new_api` as I rebased on that

File: src/jaxsim/high_level/model.py [hunk 1]
```diff
@@ -1137,25 +1137,36 @@ def forward_dynamics_crb(
...
-            pred=self.floating_base(),
-            on_true=jnp.linalg.inv(M) @ ((S @ τ) - h + J.T @ f_ext),
-            on_false=jnp.vstack(
-                [
-                    jnp.zeros(shape=(6, 1)),
-                    jnp.linalg.inv(M[6:, 6:])
-                    @ ((S @ τ)[6:] - h[6:] + J[:, 6:].T @ f_ext),
-                ]
-            ),
-        ).squeeze()
```
Comment (diegoferigo) [inline src/jaxsim/high_level/model.py:1158]: These lines were necessary if I recall to jit-compile this method and get correct results on fixed-based robots. In that case, the generalized acceleration should be computed differently from the floating-base case.

File: src/jaxsim/high_level/model.py [hunk 2]
```diff
@@ -1479,6 +1490,88 @@ def integrate(
...
+        joint_names = joint_names or self.joint_names()
+
+        if inertias.size != len(joint_names):
+            raise ValueError("Wrong arguments size", inertias.size, len(joint_names))
+
+        self.physics_model._joint_motor_inertia.update(
+            dict(zip(self.physics_model._joint_motor_inertia, inertias))
+        )
+
+        logging.info("Setting attribute `motor_inertias`")
```
Comment (diegoferigo) [inline src/jaxsim/high_level/model.py:1510]: do we want this log message always printed?

File: src/jaxsim/high_level/model.py [hunk 3]
```diff
@@ -1479,6 +1490,88 @@ def integrate(
...
+            if gr != 0 and self.motor_inertias()[idx] == 0:
+                raise ValueError(
+                    f"Zero motor inertia with non-zero gear ratio found in position {idx}"
+                )
+
+        self.physics_model._joint_motor_gear_ratio.update(
+            dict(zip(self.physics_model._joint_motor_gear_ratio, gear_ratios))
+        )
+
+        logging.info("Setting attribute `motor_gear_ratios`")
```
Comment (diegoferigo) [inline src/jaxsim/high_level/model.py:1532]: same here

File: src/jaxsim/high_level/model.py [hunk 4]
```diff
@@ -1479,6 +1490,88 @@ def integrate(
...
+                )
+            )
+        )
+
+        # self.physics_model._joint_motor_viscous_friction |= dict(
+        #     zip(
+        #         self.physics_model._joint_motor_viscous_friction.keys(),
+        #         viscous_frictions,
+        #     )
+        # )
```
Comment (diegoferigo) [inline src/jaxsim/high_level/model.py:1559]: leftover?

File: src/jaxsim/high_level/model.py [hunk 5]
```diff
@@ -1479,6 +1490,88 @@ def integrate(
...
+
+        logging.info("Setting attribute `motor_viscous_frictions`")
+
+    @functools.partial(oop.jax_tf.method_ro, jit=False)
+    def motor_inertias(self) -> jtp.Vector:
+        return jnp.array([*self.physics_model._joint_motor_inertia.values()])
```
Comment (diegoferigo) [inline src/jaxsim/high_level/model.py:1565]: let's enforce `dtype=float` also here

File: src/jaxsim/high_level/model.py [hunk 6]
```diff
@@ -1479,6 +1490,88 @@ def integrate(
...
+
+    @functools.partial(oop.jax_tf.method_ro, jit=False)
+    def motor_gear_ratios(self) -> jtp.Vector:
+        return jnp.array([*self.physics_model._joint_motor_gear_ratio.values()])
```
Comment (diegoferigo) [inline src/jaxsim/high_level/model.py:1569]: let's enforce `dtype=float` also here

File: src/jaxsim/high_level/model.py [hunk 7]
```diff
@@ -1479,6 +1490,88 @@ def integrate(
...
+
+    @functools.partial(oop.jax_tf.method_ro, jit=False)
+    def motor_viscous_frictions(self) -> jtp.Vector:
+        return jnp.array([*self.physics_model._joint_motor_viscous_friction.values()])
```
Comment (diegoferigo) [inline src/jaxsim/high_level/model.py:1573]: let's enforce `dtype=float` also here

File: src/jaxsim/physics/algos/aba.py [hunk 8]
```diff
@@ -81,27 +93,36 @@ def aba(
...
+        i_X_λi, v, c, m_v, m_c, MA, pA, pR, i_X_0 = carry
 
         # Compute parent-to-child transform
         i_X_λi_i = i_X_pre[i] @ pre_X_λi[i]
         i_X_λi = i_X_λi.at[i].set(i_X_λi_i)
 
         # Propagate link velocity
-        vJ = S[i] * qd[ii] if qd.size != 0 else S[i] * 0
+        vJ = S[i] * qd[ii] * (qd.size != 0)
+        m_vJ = m_S[i] * qd[ii] * (qd.size != 0)
```
Comment (diegoferigo) [inline src/jaxsim/physics/algos/aba.py:114]: Did you validate jit compilation? I remember I had to use the if/else statement in this way otherwise jax was complaining

File: src/jaxsim/physics/algos/aba.py [hunk 9]
```diff
@@ -115,46 +136,72 @@ def loop_body_pass1(carry: Pass1Carry, i: jtp.Int) -> Tuple[Pass1Carry, None]:
...
     def loop_body_pass2(carry: Pass2Carry, i: jtp.Int) -> Tuple[Pass2Carry, None]:
         ii = i - 1
-        U, d, u, MA, pA = carry
+        U, m_U, d, u, m_u, MA, pA = carry
 
         # Compute intermediate results
+        u_i = tau[ii] - S[i].T @ pA[i] if tau.size != 0 else -S[i].T @ pA[i]
+        u = u.at[i].set(u_i.squeeze())
+
+        has_motors = Γ[i] != 1
```
Comment (diegoferigo) [inline src/jaxsim/physics/algos/aba.py:177]: comparing floats (Γ[i] is a float) is dangerous, should we use `jnp.allclose(Γ[i], 1)`?

**Review by diegoferigo — CHANGES_REQUESTED** [review]
First round of review, it looks good even if I still have some doubts on the updated ABA and RNEA algorithms. If we want to merge into `main` and the new methods are not yet 100% tested and benchmarked, let's consider to commit them for the time being as `aba_motors.py` and `rnea_motors.py`. We can fix the remaining problems in the future without affecting current functionality.

**Comment by flferretti** [comment]
Also this has been closed once merged `new_api`. I'll reopen it with the addressed comments

File: src/jaxsim/high_level/model.py (hunk 4 above)
Comment (flferretti) [reply src/jaxsim/high_level/model.py:1559]: Yes, thanks

File: src/jaxsim/high_level/model.py (hunk 6 above)
Comment (flferretti) [reply src/jaxsim/high_level/model.py:1569]: Done in https://github.com/ami-iit/jaxsim/pull/62/commits/9d4754b5a054c42f4248348ea1af20d1079fbb6f

File: src/jaxsim/high_level/model.py (hunk 5 above)
Comment (flferretti) [reply src/jaxsim/high_level/model.py:1565]: Done in https://github.com/ami-iit/jaxsim/pull/62/commits/9d4754b5a054c42f4248348ea1af20d1079fbb6f

File: src/jaxsim/high_level/model.py (hunk 7 above)
Comment (flferretti) [reply src/jaxsim/high_level/model.py:1573]: Done in https://github.com/ami-iit/jaxsim/pull/62/commits/9d4754b5a054c42f4248348ea1af20d1079fbb6f

File: src/jaxsim/physics/algos/aba.py (hunk 8 above)
Comment (flferretti) [reply src/jaxsim/physics/algos/aba.py:114]: Yes, it did not complain, but we can further investigate on this

File: src/jaxsim/physics/algos/aba.py (hunk 9 above)
Comment (flferretti) [reply src/jaxsim/physics/algos/aba.py:177]: Done in https://github.com/ami-iit/jaxsim/pull/62/commits/48d99f3d94274d38bf8fcc436ee278738e2ca6dc

File: src/jaxsim/high_level/model.py (hunk 2 above)
Comment (flferretti) [reply src/jaxsim/high_level/model.py:1510]: With a normal installation it should not get printed as the logging level is set to by default. The log message will appear only when installing in development mode: https://github.com/ami-iit/jaxsim/blob/293161cf16f486f2a2eb13ce2e963d951b3f8820/src/jaxsim/__init__.py#L48-L51

File: src/jaxsim/high_level/model.py (hunk 3 above)
Comment (flferretti) [reply src/jaxsim/high_level/model.py:1532]: See comment above
